File: Components/DevExpressVCL/Demos/ExpressSpreadSheet/CBuilder/CustomDraw/CustomDrawDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "CustomDrawDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BaseForm"
#pragma link "cxClasses"
#pragma link "cxLookAndFeels"
#pragma link "cxControls"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "dxCore"
#pragma link "dxCoreClasses"
#pragma link "dxHashUtils"
#pragma link "dxSpreadSheet"
#pragma link "dxSpreadSheetClasses"
#pragma link "dxSpreadSheetCore"
#pragma link "dxSpreadSheetFormulas"
#pragma link "dxSpreadSheetFunctions"
#pragma link "dxSpreadSheetGraphics"
#pragma link "dxSpreadSheetTypes"
#pragma link "dxSpreadSheetFormulaBar"
#pragma link "cxSplitter"
#pragma resource "*.dfm"
TfrmCustomDraw *frmCustomDraw;
//---------------------------------------------------------------------------
__fastcall TfrmCustomDraw::TfrmCustomDraw(TComponent* Owner)
	: TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomDraw::FormCreate(TObject *Sender)
{
  TfmBaseForm::FormCreate(Sender);
  SpreadSheet->LoadFromFile("..\\..\\Data\\CalculatorOfArea.xlsx");
  WindowState = wsMaximized;
}
//---------------------------------------------------------------------------

TdxSpreadSheet* TfrmCustomDraw::GetSpreadSheet()
{
  return SpreadSheet;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomDraw::SpreadSheetCustomDrawTableViewCell(TdxSpreadSheetTableView *Sender, TcxCanvas *ACanvas, TdxSpreadSheetTableViewCellViewInfo *AViewInfo,
		  bool &AHandled)
{
  if ((AViewInfo->Cell == NULL)||(AViewInfo->Cell->DataType != cdtFormula))
	return;
  if ((AViewInfo->Cell->AsFormula->ErrorCode != ecNone)||(AViewInfo->Cell->AsFormula->ResultValue->ErrorCode != ecNone)||
	(!(VarIsNumeric(AViewInfo->Cell->AsFormula->Value))||(AViewInfo->Cell->AsFormula->Value < 0)))
  {
	ACanvas->Brush->Color = clRed;
	ACanvas->Font->Color = clAqua;
	DrawCallout(ACanvas, AViewInfo);
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmCustomDraw::DrawCallout(TcxCanvas* ACanvas, TdxSpreadSheetTableViewCellViewInfo* AViewInfo)
{
	TcxRegion *AClipRgn, *AFrameRgn;
	TPoint FramePoints[6], TL;

	TL.x = AViewInfo->Bounds.left;
	TL.y = AViewInfo->Bounds.top;
	FramePoints[0] = cxPointOffset(TL, 0, 10);
	FramePoints[1] = cxPointOffset(FramePoints[0], 5, -10);
	FramePoints[2] = cxPointOffset(FramePoints[1], 200, 0);
	FramePoints[3] = cxPointOffset(FramePoints[2], 0, -20);
	FramePoints[4] = cxPointOffset(FramePoints[3], -205, 0);
	FramePoints[5] = FramePoints[0];

	AFrameRgn = new TcxRegion();
	AFrameRgn->Handle = CreatePolygonRgn(FramePoints, 6, WINDING);
	ACanvas->SaveState();
	AClipRgn = new TcxRegion(AViewInfo->ViewInfo->CellsArea);
	AClipRgn->Combine(AFrameRgn, roIntersect, false);
	ACanvas->SetClipRegion(AClipRgn, roSet, false);
	FillRegionByColor(ACanvas->Handle, AFrameRgn->Handle, clInfoBk);
	FrameRgn(ACanvas->Handle, AFrameRgn->Handle, static_cast<HBRUSH__*>(GetStockObject(BLACK_BRUSH)), 1, 1);
	ACanvas->Font->Name = "Tahoma";
	ACanvas->Font->Size = 10;
	ACanvas->Font->Color = clBlack;
	ACanvas->Font->Style = TFontStyles() << fsBold;
	ACanvas->Brush->Style = bsClear;
	ACanvas->DrawTexT("Incorrect parameter values", cxRect(FramePoints[4], FramePoints[2]), cxAlignVCenter || cxAlignHCenter || cxSingleLine);
	ACanvas->RestoreState();
	ACanvas->SetClipRegion(AClipRgn, roSubtract);
	AFrameRgn->Free();
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressSpreadSheet/Delphi/CustomDraw/CustomDrawDemoMain.pas
unit CustomDrawDemoMain;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms, cxGeometry,
  Dialogs, cxGraphics, cxControls, cxLookAndFeels, cxLookAndFeelPainters, dxSpreadSheetCore, dxSpreadSheetFunctions,
  dxSpreadSheetGraphics, dxSpreadSheetClasses, dxSpreadSheetTypes, dxSpreadSheet, BaseForm, dxCore, dxCoreClasses,
  dxHashUtils, dxSpreadSheetCoreHistory, dxSpreadSheetCoreStyles, dxSpreadSheetCoreStrs,
  dxSpreadSheetConditionalFormatting, dxSpreadSheetConditionalFormattingRules, dxSpreadSheetContainers,
  dxSpreadSheetFormulas, dxSpreadSheetHyperlinks, dxSpreadSheetStyles, dxSpreadSheetPrinting, dxSpreadSheetUtils,
  cxSplitter, dxSpreadSheetFormulaBar;

type
  TfrmCustomDraw = class(TfmBaseForm)
    SpreadSheet: TdxSpreadSheet;
    FormulaBar: TdxSpreadSheetFormulaBar;
    FormulaBarSplitter: TcxSplitter;
    procedure FormCreate(Sender: TObject);
    procedure SpreadSheetCustomDrawTableViewCell(Sender: TdxSpreadSheetTableView; ACanvas: TcxCanvas;
      AViewInfo: TdxSpreadSheetTableViewCellViewInfo; var AHandled: Boolean);
  private
  protected
    function GetSpreadSheet: TdxSpreadSheet; override;
  public
    procedure DrawCallout(ACanvas: TcxCanvas; AViewInfo: TdxSpreadSheetTableViewCellViewInfo);
  end;

var
  frmCustomDraw: TfrmCustomDraw;

implementation

{$R *.dfm}

uses
  Types;

procedure TfrmCustomDraw.FormCreate(Sender: TObject);
begin
  SpreadSheet.LoadFromFile('..\..\Data\CalculatorOfArea.xlsx');
end;

function TfrmCustomDraw.GetSpreadSheet: TdxSpreadSheet;
begin
  Result := SpreadSheet;
end;

procedure TfrmCustomDraw.DrawCallout(ACanvas: TcxCanvas; AViewInfo: TdxSpreadSheetTableViewCellViewInfo);
var
  AClipRgn, AFrameRgn: TcxRegion;
  FramePoints: array[0..5] of TPoint;
begin
  FramePoints[0] := cxPointOffset(AViewInfo.Bounds.TopLeft, 0, 10);
  FramePoints[1] := cxPointOffset(FramePoints[0], 5, -10);
  FramePoints[2] := cxPointOffset(FramePoints[1], 200, 0);
  FramePoints[3] := cxPointOffset(FramePoints[2], 0, -20);
  FramePoints[4] := cxPointOffset(FramePoints[3], -205, 0);
  FramePoints[5] := FramePoints[0];
  AFrameRgn := TcxRegion.Create;
  AFrameRgn.Handle := CreatePolygonRgn(FramePoints, Length(FramePoints), WINDING);
  ACanvas.SaveState;
  AClipRgn := TcxRegion.Create(AViewInfo.ViewInfo.CellsArea);
  AClipRgn.Combine(AFrameRgn, roIntersect, False);
  ACanvas.SetClipRegion(AClipRgn, roSet, False);
  FillRegionByColor(ACanvas.Handle, AFrameRgn.Handle, clInfoBk);
  FrameRgn(ACanvas.Handle, AFrameRgn.Handle, GetStockObject(BLACK_BRUSH), 1, 1);
  ACanvas.Font.Name := 'Tahoma';
  ACanvas.Font.Size := 10;
  ACanvas.Font.Color := clBlack;
  ACanvas.Font.Style := [fsBold];
  ACanvas.Brush.Style := bsClear;
  ACanvas.DrawTexT('Incorrect parameter values', cxRect(FramePoints[4], FramePoints[2]),
    cxAlignVCenter or cxAlignHCenter or cxSingleLine);
  ACanvas.RestoreState;
  ACanvas.SetClipRegion(AClipRgn, roSubtract);
  AFrameRgn.Free;
end;

procedure TfrmCustomDraw.SpreadSheetCustomDrawTableViewCell(
  Sender: TdxSpreadSheetTableView; ACanvas: TcxCanvas;
  AViewInfo: TdxSpreadSheetTableViewCellViewInfo; var AHandled: Boolean);
begin
  if (AViewInfo.Cell = nil) or  (AViewInfo.Cell.DataType <> cdtFormula) then Exit;
  if  (AViewInfo.Cell.AsFormula.ErrorCode <> ecNone) or (AViewInfo.Cell.AsFormula.ResultValue.ErrorCode <> ecNone) or
    (not VarIsNumeric(AViewInfo.Cell.AsFormula.Value) or (AViewInfo.Cell.AsFormula.Value < 0)) then
  begin
    ACanvas.Brush.Color := clRed;
    ACanvas.Font.Color := clAqua;
    DrawCallout(ACanvas, AViewInfo);
  end;
end;

end.
